#2346 Fix problem with downstream URLs where query parameter names get messed up if they contain a placeholder that has a non-empty value

bhargav-polara opened the pull request:
## Fixes #2346 
- #2346 

### Background

After upgrading to Ocelot [23.4.3](https://github.com/ThreeMammals/Ocelot/releases/tag/23.4.3), query parameters whose names contain the substring `id` (e.g., `customer_id`) are incorrectly rewritten or corrupted if another route exists with a `{id}` path parameter. See #2346 for details and repro steps.

### Problem

When Ocelot tries to remove consumed query string parameters from the downstream request, it was using a simple string `Contains` check. This caused substring matches (like the `id` in `customer_id`) to be incorrectly interpreted as route tokens—resulting in corruption (e.g., `customer_id=5172980` becoming `customer5172980`).

### Solution

**Changed logic in `RemoveQueryStringParametersThatHaveBeenUsedInTemplate` to use a Regex match instead of string `Contains`.**  
This ensures only exact parameter matches are identified and removed, and substrings like `id` in `customer_id` are preserved correctly.

```csharp
if (!Regex.IsMatch(
        downstreamRequest.Query,
        $@"(?<=^|\?|&){Regex.Escape(parameter)}",
        RegexOptions.IgnoreCase))
{
    continue;
}
```

---

### How to Reproduce

See steps in #2346:

1. Define two routes:
    - One with `/finance/v1/payment-methods/{id}`
    - One with `/finance/v1/payment-methods`
2. Call GET `/finance/v1/payment-methods?customer_id=5172980`
3. The downstream request should forward the query string unmodified.

### Testing

- Manually tested scenarios from [2346](https://github.com/ThreeMammals/Ocelot/issues/2346); parameters like `customer_id=5172980` are now correctly forwarded downstream without corruption.
- All existing unit and integration tests pass (`dotnet test`).

### Checklist
- [x] Branch based on `develop`
- [x] Code follows contribution guidelines
- [x] Tests passing

bhargav-polara committed:
Update DownstreamUrlCreatorMiddleware.cs

bhargav-polara committed:
Update DownstreamUrlCreatorMiddlewareTests.cs

raman-m commented:
> ### Problem
> When Ocelot tries to remove consumed query string parameters from the downstream request, **it was using a simple string `Contains` check. This caused substring matches (like the `id` in `customer_id`) to be incorrectly interpreted as route tokens**—resulting in corruption (e.g., `customer_id=5172980` becoming `customer5172980`).
> 

Right! It was my fault when releasing the Ocelot [23.4.3](https://github.com/ThreeMammals/Ocelot/releases/tag/23.4.3) patch. Sometimes, when you solve a big problem, there’s a chance you might introduce a small, unexpected one 👶

My idea to use the `Contains` method works/worked well when we also check that both names are the same length. This means the `Contains` method requires both name lengths to be equal. However, we still need to parse the parameter in the `downstreamRequest.Query` string or verify its existence in some way. I think the easiest solution is to reuse the functionality of the `HttpContext.Request.Query` object 👇 
```csharp
IQueryCollection query = context.Request.Query;
bool exists = query.ContainsKey(name) && query.TryGetValue(name, out var pvalue) && pvalue == nAndV.Value;
if (!exists) // !downstreamRequest.Query.Contains(parameter)
{
    continue;
}
```
To do that, we just need an additional `context` argument, and that’s all. Parsing the parameter value doesn’t seem necessary, though it might depend. The algorithm following these lines could also be optimized.
https://github.com/ThreeMammals/Ocelot/blob/9fc4e78d314ad97d836b35c9b9bb188dba49d556/src/Ocelot/DownstreamUrlCreator/DownstreamUrlCreatorMiddleware.cs#L135

raman-m commented on src/Ocelot/DownstreamUrlCreator/DownstreamUrlCreatorMiddleware.cs:
```diff
@@ -132,7 +132,10 @@ protected static void RemoveQueryStringParametersThatHaveBeenUsedInTemplate(Down
+            if (!Regex.IsMatch(
+                    downstreamRequest.Query,
+                    $@"(?<=^|\?|&){Regex.Escape(parameter)}",
+                    RegexOptions.IgnoreCase))
```
The original idea in the [23.4.3](https://github.com/ThreeMammals/Ocelot/releases/tag/23.4.3) patch was to reduce the number of `Regex` objects. However, the non-cached version of the regular expression builds the `Regex` on the fly and then reuses it for matching. This is a resource-intensive algo that consumes both CPU and RAM. From what I understand, your solution uses a `Regex` to ensure there are no `=`, `?`, or `&` characters in the full query string before the parameter. In the end, I found this approach to be costly in terms of CPU & RAM usage.

And here’s the dilemma: should we optimize `Regex` usage by adding a cache, or explore [other potential Regex-free solutions](https://github.com/ThreeMammals/Ocelot/pull/2351#issuecomment-3703775650) instead?

raman-m commented on test/Ocelot.UnitTests/DownstreamUrlCreator/DownstreamUrlCreatorMiddlewareTests.cs:
```diff
@@ -662,6 +662,37 @@ public async Task ShouldNotFailToHandleUrlWithSpecialRegexChars(string urlPath)
 
+    [Fact]
+    [Trait("Bug", "2346")]
+    public async Task Should_not_corrupt_query_parameter_names_containing_id_when_route_has_id_placeholder()
```
Having a single unit test is fine, but what about covering more edge cases from the original issue and perhaps adding even more scenarios?
Converting to a `Theory`?

raman-m reviewed (changes requested):
While this solution should fix the bug, it looks like we still need more thorough unit testing with different URLs. Additionally, the lack of acceptance tests is preventing me from approving this version of the bugfix.
Let’s work on finding the optimal solution, as this is a critical middleware for the routing feature. I recommend spending more time refining the approach.

bhargav-polara committed:
Added more test Cases and refactor RemoveQueryStringParametersThatHaveBeenUsedInTemplate Logic

bhargav-polara commented:
@raman-m I had tried out [this ](https://github.com/ThreeMammals/Ocelot/pull/2351#issuecomment-3703775650) your suggestion but after applying other testcases are getting failed. I have refactored code and added more test Cases.

RaynaldM commented on src/Ocelot/DownstreamUrlCreator/DownstreamUrlCreatorMiddleware.cs:
```diff
@@ -126,27 +127,64 @@ protected static string MergeQueryStringsWithoutDuplicateValues(string queryStri
+                if (placeholderLookup.TryGetValue(queryKey, out var placeholderValue))
+                {
+                    // Remove only the values that match the placeholder value exactly
+                    var filteredValues = queryValues.Where(qv => !string.Equals(qv, placeholderValue, StringComparison.Ordinal)).ToList();
```
```suggestion
                    var filteredValues = queryValues.Where(qv => !string.Equals(qv, placeholderValue, StringComparison.Ordinal)).ToArray();
```

raman-m replied:
@RaynaldM 
You approved the PR even though commit https://github.com/ThreeMammals/Ocelot/pull/2351/commits/21661a4eec7a53e32704de33987a95ce42ee6d1a [tests failed](https://github.com/ThreeMammals/Ocelot/actions/runs/20706286947/job/59455855467), seriously?..
The author's refactoring in https://github.com/ThreeMammals/Ocelot/pull/2351/commits/21661a4eec7a53e32704de33987a95ce42ee6d1a wasn't necessary, to be fair. Instead of taking [my suggestions](https://github.com/ThreeMammals/Ocelot/pull/2351#pullrequestreview-3621897853) with recommendation to resolve the issues, he went for [aggressive algo refactoring](https://github.com/ThreeMammals/Ocelot/pull/2351/commits/21661a4eec7a53e32704de33987a95ce42ee6d1a#diff-080d008993a0eca45a728e6499876a947129ddb8ae8e3310422f2c0dfc19a7e8R129-R130) without checking how the tests ran locally.
### If you are okay with approving such a contribution, then I am at a loss for words.

raman-m commented on src/Ocelot/DownstreamUrlCreator/DownstreamUrlCreatorMiddleware.cs:
```diff
@@ -126,27 +127,64 @@ protected static string MergeQueryStringsWithoutDuplicateValues(string queryStri
-                continue;
+                rawQuery =
+                    rawQuery.Substring(0, firstQuestionMark + 1) +
+                    rawQuery.Substring(firstQuestionMark + 1).Replace("?", "&");
```
`// Normalize legacy malformed query (?a=1?b=2)` Wow!
Malformed query strings can’t occur in this middleware because `HttpContext.Request` is already initialized and the URL is well-formed, including the `DownstreamRequest.Query` property which is preinitialized by the following code:
https://github.com/ThreeMammals/Ocelot/blob/9fc4e78d314ad97d836b35c9b9bb188dba49d556/src/Ocelot/Request/Middleware/DownstreamRequest.cs#L20
The `_request` field is an `HttpRequestMessage` that's being created here:
https://github.com/ThreeMammals/Ocelot/blob/9fc4e78d314ad97d836b35c9b9bb188dba49d556/src/Ocelot/Request/Mapper/RequestMapper.cs#L15-L22
And we should focus our attention on this method:
https://github.com/ThreeMammals/Ocelot/blob/9fc4e78d314ad97d836b35c9b9bb188dba49d556/src/Ocelot/Request/Mapper/RequestMapper.cs#L76
Thus, `DownstreamRequest.Query` is simply encoded `HttpRequest.QueryString` and/or `HttpRequest.Query`.

An acceptance test should be created to confirm this is possible, and if the test is developed and proves it, the code block can stay.

raman-m commented on src/Ocelot/DownstreamUrlCreator/DownstreamUrlCreatorMiddleware.cs:
```diff
@@ -126,27 +127,64 @@ protected static string MergeQueryStringsWithoutDuplicateValues(string queryStri
-            downstreamRequest.Query = builder.Length > 0
-                ? builder.Remove(0, 1).Insert(0, QuestionMark).ToString()
+            // Parse the NORMALIZED query
+            var parsedQuery = QueryHelpers.ParseQuery(rawQuery);
```
Helpful method, but I am not sure we need to parse the string at all. We just need to check if the query parameter exists. I'll keep this helper in mind if it becomes necessary.

raman-m commented on src/Ocelot/DownstreamUrlCreator/DownstreamUrlCreatorMiddleware.cs:
```diff
@@ -126,27 +127,64 @@ protected static string MergeQueryStringsWithoutDuplicateValues(string queryStri
+            
+            // Create a case-insensitive lookup for template placeholders
+            var placeholderLookup = templatePlaceholders
+                .ToDictionary(p => p.Name.Trim(OpeningBrace, ClosingBrace), p => p.Value, StringComparer.OrdinalIgnoreCase);
```
Consider refactoring `PlaceholderNameAndValue` instead of trimming the values. Also, why use `ToDictionary`? It adds unnecessary pressure on RAM and the garbage collector by creating a new object on the heap!

raman-m commented on src/Ocelot/DownstreamUrlCreator/DownstreamUrlCreatorMiddleware.cs:
```diff
@@ -126,27 +127,64 @@ protected static string MergeQueryStringsWithoutDuplicateValues(string queryStri
+            // Build new query string by filtering out matching placeholders
+            var remainingParameters = new List<string>();
+            
+            foreach (var kvp in parsedQuery)
```
This is an iteration loop through parameters, not the collection of placeholders.
The question is: **why?** 😄

raman-m commented on src/Ocelot/DownstreamUrlCreator/DownstreamUrlCreatorMiddleware.cs:
```diff
@@ -126,27 +127,64 @@ protected static string MergeQueryStringsWithoutDuplicateValues(string queryStri
+                    foreach (var value in queryValues)
+                    {
+                        remainingParameters.Add($"{queryKey}={HttpUtility.UrlEncode(value)}");
+                    }
```
If you already have a parsed version of the query, like `var parsedQuery = QueryHelpers.ParseQuery(rawQuery)`, why create a new collection? 🤨
Do you think a dictionary doesn’t have a remove method, like `parsedQuery.Remove(name)`? 🤣 That is hilarious!

raman-m reviewed (changes requested):
Yahoo! The rodeo goes on! 🤠 
Did you notice the CI build [failed](https://github.com/ThreeMammals/Ocelot/actions/runs/20706286947/job/59455855467)? No rush, though I'm planning to revert the last commit https://github.com/ThreeMammals/Ocelot/pull/2351/commits/21661a4eec7a53e32704de33987a95ce42ee6d1a.

You're a talented vibe-coder, and the new algorithm seems solid and will handle query strings properly, but instead of the old 10 lines, we now have 30 lines of code proposed by the AI coding agent. This version will heavily consume RAM due to two new collections in the algorithm. Honestly, I prefer the previous `Regex`-based version since it's better to use more CPU cycles than to increase RAM usage. Have you seen how expensive RAM modules are these days? 😉

> I had tried out https://github.com/ThreeMammals/Ocelot/pull/2351#issuecomment-3703775650your suggestion but after applying other testcases are getting failed. **I have refactored code and added more test Cases.**

You've added new unit tests generated by the AI coding agent but wrote no acceptance tests. And you didn't run the tests locally, so now the old Routing acceptance tests are failing 👇 
<img width="678" height="372" alt="image" src="https://github.com/user-attachments/assets/f9d43869-daf8-4042-8b0a-a9419d233558" />

Sorry, the last commit will be reverted because the old `Regex` version is more suitable.

raman-m committed:
Revert "Added more test Cases and refactor RemoveQueryStringParametersThatHaveBeenUsedInTemplate Logic"

This reverts commit 21661a4eec7a53e32704de33987a95ce42ee6d1a.

bhargav-polara committed:
Replaced the regex-based logic in RemoveQueryStringParametersThatHaveBeenUsedInTemplate

bhargav-polara committed:
Revert "Replaced the regex-based logic in RemoveQueryStringParametersThatHaveBeenUsedInTemplate"

This reverts commit 40358588bc995ab483b5963738bfaeafb09deab4.

bhargav-polara committed:
Replaced the regex-based logic in RemoveQueryStringParametersThatHaveBeenUsedInTemplate

bhargav-polara commented:
@raman-m I have refactored the code based on your suggestions and also made a few additional changes to address the failing test cases that appeared after applying them.

raman-m commented:
It is great if the unit tests and old acceptance tests are now passing.
Consider adding new acceptance tests to the [RoutingWithQueryStringTests](https://github.com/ThreeMammals/Ocelot/blob/develop/test/Ocelot.AcceptanceTests/Routing/RoutingWithQueryStringTests.cs) class. Be sure to include the test case for #2346 to confirm the bug fix works properly.
Please note, I will perform the final code review after writing the acceptance tests.

bhargav-polara committed:
Update RoutingWithQueryStringTests.cs

bhargav-polara commented:
@raman-m Added new acceptance tests

raman-m commented:
### Ready for code review ✔️

raman-m committed:
Testing the "Catch All Query String" feat to make sure there is no regression

raman-m committed:
Refactor algorithm to take full advantage of QueryHelpers

raman-m committed:
Code review by @raman-m ~ improved testing

raman-m committed:
Roll back the status badge targets to develop

raman-m reviewed (approved):
## Ready for delivery ✅ 
- Code review ✔️ 
- Unit testing ✔️ ✔️ 
- Acceptance testing ✔️ ✔️
